File: backend/routers/tasks.py
"""Tasks API endpoints for background agent task management."""
import asyncio
import json
import logging
from typing import Optional

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import JSONResponse, StreamingResponse

from schemas.task import TaskCreate, TaskResponse, TaskMessageRequest, RunningTaskCount
from core.task_manager import task_manager
from core.skill_manager import skill_manager
from database import db

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get("", response_model=list[TaskResponse])
async def list_tasks(
    status: Optional[str] = Query(None, description="Filter by status (comma-separated, OR semantics)"),
    agent_id: Optional[str] = Query(None, description="Filter by agent ID"),
    workspace_id: Optional[str] = Query(None, description="Filter by workspace ID"),
    completed_after: Optional[str] = Query(None, description="Filter completed tasks after ISO 8601 date"),
):
    """List all tasks, optionally filtered.

    Status filter uses comma-separated values with OR semantics:
    ``?status=wip,draft,blocked`` means (status=wip OR status=draft OR status=blocked).
    Different parameter types use AND semantics:
    ``?status=completed&workspace_id=abc`` means (status=completed AND workspace_id=abc).
    """
    tasks = await task_manager.list_tasks(
        status=status,
        agent_id=agent_id,
        workspace_id=workspace_id,
        completed_after=completed_after,
    )
    return [TaskResponse(**task) for task in tasks]


@router.get("/running/count", response_model=RunningTaskCount)
async def get_running_count():
    """Get count of running tasks (for sidebar badge)."""
    count = await task_manager.get_running_count()
    return RunningTaskCount(count=count)


@router.get("/{task_id}", response_model=TaskResponse)
async def get_task(task_id: str):
    """Get a specific task by ID."""
    task = await task_manager.get_task(task_id)
    if not task:
        raise HTTPException(status_code=404, detail=f"Task {task_id} not found")
    return TaskResponse(**task)


@router.post("", response_model=TaskResponse)
async def create_task(request: TaskCreate):
    """Create and start a new background task.

    Validates workspace policy before execution. If required skills or MCPs
    are disabled in the workspace, returns 409 Conflict with policy_violations.

    Requirements: 26.1-26.7, 34.1-34.7
    """
    # --- Policy enforcement: check required_skills / required_mcps ---
    if request.required_skills or request.required_mcps:
        violations = []
        ws_id = request.workspace_id

        # Resolve default workspace if not provided
        if not ws_id:
            config = await db.workspace_config.get_config()
            ws_id = config["id"] if config else None

        if ws_id:
            # Check required skills — in filesystem model all skills are always
            # enabled, so we only verify the skill exists in the cache.
            skill_cache = await skill_manager.get_cache()
            for skill_id in (request.required_skills or []):
                if skill_id not in skill_cache:
                    violations.append({
                        "entity_type": "skill",
                        "entity_id": skill_id,
                        "message": f"Skill {skill_id} not found in workspace {ws_id}",
                        "suggestedAction": f"Install skill {skill_id} or check the name",
                    })

            # Check required MCPs
            for mcp_id in (request.required_mcps or []):
                row = await db.workspace_mcps.get_by_workspace_and_mcp(ws_id, mcp_id)
                if not row or not row.get("enabled", 1):
                    violations.append({
                        "entity_type": "mcp",
                        "entity_id": mcp_id,
                        "message": f"MCP {mcp_id} is not enabled in workspace {ws_id}",
                        "suggestedAction": f"Enable MCP {mcp_id} in workspace settings",
                    })

        if violations:
            return JSONResponse(
                status_code=409,
                content={
                    "code": "POLICY_VIOLATION",
                    "message": "Required capabilities are not enabled in the workspace",
                    "policy_violations": violations,
                    "suggested_action": "Enable the required capabilities in workspace settings",
                },
            )

    try:
        task = await task_manager.create_task(
            agent_id=request.agent_id,
            message=request.message,
            content=request.content,
            enable_skills=request.enable_skills,
            enable_mcp=request.enable_mcp,
            workspace_id=request.workspace_id,
            source_todo_id=request.source_todo_id,
            priority=request.priority,
            description=request.description,
        )
        return TaskResponse(**task)
    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{task_id}")
async def delete_task(task_id: str):
    """Delete a task (cancels if running)."""
    deleted = await task_manager.delete_task(task_id)
    if not deleted:
        raise HTTPException(status_code=404, detail=f"Task {task_id} not found")
    return {"status": "deleted", "task_id": task_id}


@router.post("/{task_id}/cancel")
async def cancel_task(task_id: str):
    """Cancel a running task."""
    cancelled = await task_manager.cancel_task(task_id)
    if not cancelled:
        raise HTTPException(status_code=400, detail=f"Task {task_id} is not running")
    return {"status": "cancelled", "task_id": task_id}


@router.post("/{task_id}/message")
async def send_message(task_id: str, request: TaskMessageRequest):
    """Send a message to a running task."""
    success = await task_manager.send_message(
        task_id=task_id,
        message=request.message,
        content=request.content,
    )
    if not success:
        raise HTTPException(status_code=400, detail=f"Task {task_id} is not running or not accepting messages")
    return {"status": "sent", "task_id": task_id}


@router.get("/{task_id}/stream")
async def stream_task(task_id: str):
    """Subscribe to task events via SSE."""
    task = await task_manager.get_task(task_id)
    if not task:
        raise HTTPException(status_code=404, detail=f"Task {task_id} not found")

    async def event_generator():
        try:
            async for event in task_manager.subscribe(task_id):
                yield f"data: {json.dumps(event)}\n\n"
        except asyncio.CancelledError:
            logger.info("SSE stream cancelled for task %s", task_id)

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: desktop/src/services/tasks.ts
/**
 * Tasks service for background agent task management.
 */
import api, { isApiError } from './api';
import { getBackendPort } from './tauri';
import type { Task, TaskCreateRequest, TaskMessageRequest, TaskStatus, PolicyViolationDetail } from '../types';
import { ErrorCodes } from '../types';

// Convert snake_case to camelCase
function toCamelCase(task: Record<string, unknown>): Task {
  return {
    id: task.id as string,
    workspaceId: (task.workspace_id as string) ?? null,
    agentId: task.agent_id as string,
    sessionId: task.session_id as string | null,
    status: task.status as TaskStatus,
    title: task.title as string,
    description: (task.description as string) ?? null,
    priority: (task.priority as string) ?? null,
    sourceTodoId: (task.source_todo_id as string) ?? null,
    blockedReason: (task.blocked_reason as string) ?? null,
    model: task.model as string | null,
    createdAt: task.created_at as string,
    startedAt: task.started_at as string | null,
    completedAt: task.completed_at as string | null,
    error: task.error as string | null,
    workDir: task.work_dir as string | null,
    reviewRequired: (task.review_required as boolean) ?? false,
    reviewRiskLevel: (task.review_risk_level as string) ?? null,
  };
}

// Convert camelCase to snake_case for requests
function toSnakeCase(request: TaskCreateRequest): Record<string, unknown> {
  return {
    agent_id: request.agentId,
    message: request.message,
    content: request.content,
    enable_skills: request.enableSkills,
    enable_mcp: request.enableMcp,
    add_dirs: request.addDirs,
  };
}

/**
 * Error thrown when task creation is blocked by workspace policy.
 * Contains the list of policy violations for UI display.
 */
export class PolicyViolationError extends Error {
  public readonly violations: PolicyViolationDetail[];
  public readonly suggestedAction: string;

  constructor(message: string, violations: PolicyViolationDetail[], suggestedAction: string) {
    super(message);
    this.name = 'PolicyViolationError';
    this.violations = violations;
    this.suggestedAction = suggestedAction;
  }
}

export const tasksService = {
  /**
   * List all tasks, optionally filtered by status, agent ID, or workspace ID.
   */
  async list(status?: TaskStatus, agentId?: string, workspaceId?: string): Promise<Task[]> {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    if (agentId) params.append('agent_id', agentId);
    if (workspaceId) params.append('workspace_id', workspaceId);

    const queryString = params.toString();
    const url = queryString ? `/tasks?${queryString}` : '/tasks';

    const response = await api.get(url);
    return response.data.map(toCamelCase);
  },

  /**
   * Get a specific task by ID.
   */
  async get(taskId: string): Promise<Task> {
    const response = await api.get(`/tasks/${taskId}`);
    return toCamelCase(response.data);
  },

  /**
   * Create and start a new background task.
   * Throws PolicyViolationError if workspace policy blocks execution (409).
   */
  async create(request: TaskCreateRequest): Promise<Task> {
    try {
      const response = await api.post('/tasks', toSnakeCase(request));
      return toCamelCase(response.data);
    } catch (error) {
      if (isApiError(error) && error.statusCode === 409 && error.code === ErrorCodes.POLICY_VIOLATION) {
        const violations = error.policyViolations ?? [];
        throw new PolicyViolationError(
          error.response.message,
          violations.map(v => ({
            entityType: (v as Record<string, string>).entity_type ?? v.entityType,
            entityId: (v as Record<string, string>).entity_id ?? v.entityId,
            message: v.message,
            suggestedAction: (v as Record<string, string>).suggested_action ?? v.suggestedAction ?? (v as Record<string, string>).suggestedAction,
          })),
          error.suggestedAction ?? 'Enable required capabilities in workspace settings',
        );
      }
      throw error;
    }
  },

  /**
   * Delete a task (cancels if running).
   */
  async delete(taskId: string): Promise<void> {
    await api.delete(`/tasks/${taskId}`);
  },

  /**
   * Cancel a running task.
   */
  async cancel(taskId: string): Promise<void> {
    await api.post(`/tasks/${taskId}/cancel`);
  },

  /**
   * Send a message to a running task.
   */
  async sendMessage(taskId: string, request: TaskMessageRequest): Promise<void> {
    await api.post(`/tasks/${taskId}/message`, {
      message: request.message,
      content: request.content,
    });
  },

  /**
   * Get count of running tasks (for sidebar badge).
   */
  async getRunningCount(): Promise<number> {
    const response = await api.get('/tasks/running/count');
    return response.data.count;
  },

  /**
   * Get SSE stream URL for a task.
   */
  getStreamUrl(taskId: string): string {
    const port = getBackendPort();
    return `http://localhost:${port}/api/tasks/${taskId}/stream`;
  },
};
